packages/server-wallet/src/wallet/index.ts

```diff
@@ -210,47 +224,52 @@ const takeActions = async (channels: Bytes32[]): Promise<ExecutionResult> => {
   while (channels.length && !error) {
-    const tx = await knex.transaction();
-
-    const setError = async (e: Error): Promise<void> => {
-      error = e;
-      await tx.rollback();
-    };
-    const markChannelAsDone = async (): Promise<any> => {
-      channels.shift() as string;
-    };
-
-    const handleAction = async (action: ProtocolAction): Promise<any> => {
-      switch (action.type) {
-        case 'SignState': {
-          const {outgoing, channelResult} = await Store.signState(action.channelId, action, tx);
-          outgoing.map(n => outbox.push(n.notice));
-          channelResults.push(channelResult);
-          return;
+    await knex.transaction(async tx => {
+      const setError = async (e: Error): Promise<void> => {
+        error = e;
+        await tx.rollback(error);
+      };
+      const markChannelAsDone = async (): Promise<any> => {
+        channels.shift() as string;
+      };
+
+      const handleAction = async (action: ProtocolAction): Promise<any> => {
+        switch (action.type) {
+          case 'SignState': {
+            const {outgoing, channelResult} = await Store.signState(action.channelId, action, tx);
+            outgoing.map(n => outbox.push(n.notice));
+            channelResults.push(channelResult);
+            return;
+          }
+          default:
+            throw 'Unimplemented';
         }
-        default:
-          throw 'Unimplemented';
-      }
-    };
+      };
 
-    // For the moment, we are only considering directly funded app channels.
-    // Thus, we can directly fetch the channel record, and immediately construct the protocol state from it.
-    // In the future, we can have an App model which collects all the relevant channels for an app channel,
-    // and a Ledger model which stores ledger-specific data (eg. queued requests)
-    const app = await Channel.forId(channels[0], undefined);
+      // For the moment, we are only considering directly funded app channels.
+      // Thus, we can directly fetch the channel record, and immediately construct the protocol state from it.
+      // In the future, we can have an App model which collects all the relevant channels for an app channel,
+      // and a Ledger model which stores ledger-specific data (eg. queued requests)
+      const app = await Store.getChannel(channels[0], tx);
 
-    try {
-      const nextAction = await Application.protocol({app: app.protocolState});
-      // TODO: handleAction might also throw an error.
-      // It would be nice for handleAction to return an Either type, pipe the right values,
-      // and handle the left values with setError
-      await Either.fold(setError, Option.fold(markChannelAsDone, handleAction))(nextAction);
-      await tx.commit();
-    } catch (err) {
-      // FIXME
-      logger.error({err}, 'Error handling action');
-      await setError(err);
-    }
+      if (!app) {
+        setError(new Error('Channel not found'));
+
+        throw 'unreachable';
+      }
```

Comment:
I'm a bit confused by this block, why does it exist?

Reply:
You should handle the case if there is no app channel. This block sets the error, and calls `tx.rollback()`, which ends up throwing an error as well.

However, typescript doesn't know that, and if you don't either throw or return, then the compiler thinks `app` may still be undefined when you leave this block.

The reality is that we should _return_  an error, and then this throw is unnecessary.

Reply:
This would indicate a corrupt DB essentially right?

Reply:
What do you mean by "indicate"? Do you mean "result in"?

Reply:
If `!app` but the wallet is executing this code, it means that some internal bit of code was run inside the wallet that thought an app existed (i.e., somehow any checks for existence of a `Channel` in the DB passed, or no check was done), and then passed it to this internal function. So... if `!app` then it's either a bug in the code, or the DB was corrupted during execution?

Maybe this is a pointless question. Just reasoning about why this code exists and we can't just assume `app` or somehow have a guarantee before the function executes.

Reply:
`takeActions` is a generic function that accepts arbitrary channel IDs, so it should gracefully handle the case where the channel you're given doesn't exist.

Something, somewhere, has to convince the typescript compiler that app is not undefined at a point where you access properties of it. That is the purpose of this block of code.

We can also assert that `app` is not undefined?